src/parallel.jl

```diff
@@ -0,0 +1,143 @@
+using Base.Threads
+using Base.Threads: @spawn, fetch
+
+
+struct Partition
+    nt::Int
+    n::Int
+    k::Int
+end
+Partition(nt,n) = Partition(nt, n, div(n, nt))
+Base.length(pt::Partition) = pt.nt
+(pt::Partition)(i) = div((i-1), pt.k) + 1, i - pt.k*div((i-1), pt.k)
+(pt::Partition)(q1, q2) = (q1-1)*pt.k + q2 
+
+each(pt::Partition) = 1:pt.nt
+
+function parallel_innermost!(partition, G, G1, G2, ∇ϕ, i, t, x, θ, t′, Q, c, b, t_old,
+    F::Union{ZigZag,FactBoomerang}, (factor, adapt), args...)
+    t, x, θ = smove_forward!(G, i, t, x, θ, t′, F)
+    ∇ϕi = ∇ϕ(x, i, args...)
+    l, lb = sλ(∇ϕi, i, x, θ, F), sλ̄(b[i], t[i] - t_old[i])
+    num = 1
+    if rand()*lb < l
+        if l >= lb
+            !adapt && error("Tuning parameter `c` too small.")
+            adapt!(c, i, factor)
+        end
+        t, x, θ = smove_forward!(G2, i, t, x, θ, t′, F)
+        θ = reflect!(i, ∇ϕi, x, θ, F)
+        for j in neighbours(G1, i)
+            b[j] = ab(G1, j, x, θ, c, F)
+            t_old[j] = t[j]
+            q1, q2 = partition(j)
+            Q[q1][q2] = t[j] + poisson_time(b[j], rand())
+        end
+        return true
+    else
+        b[i] = ab(G1, i, x, θ, c, F)
+        t_old[i] = t[i]
+        q1, q2 = partition(i)
+        Q[q1][q2] = t[i] + poisson_time(b[i], rand())
+        return false
+    end
+end
+
+function parallel_spdmp_inner!(events, partition, ti, inner, G, G1, G2, ∇ϕ, t, x, θ, Q, c, b, t_old,
+    F::Union{ZigZag,FactBoomerang}, (factor, adapt), args...)
+   n = length(x)
+   acc = num = 0
+   while true
+        num += 1
+        ii, t′ = peek(Q[ti])
+        i = partition(ti, ii)
+        if !inner[i] # need neighbours at t′
+           return false, event(i, t, x, θ, F), i, t′, acc, num
+        end
+
+        success = parallel_innermost!(partition, G, G1, G2, ∇ϕ, i, t, x, θ, t′, Q, c, b, t_old, F, (factor, adapt), args...)
+
+        success || continue
+        acc += 1
+        #return true, event(i, t, x, θ, F), i, t′, acc, num
+        push!(events[ti], event(i, t, x, θ, F))
+   end
+end
+
+function parallel_spdmp(partition, ∇ϕ, t0, x0, θ0, T, c, G, F::Union{ZigZag,FactBoomerang}, args...;
+    factor=1.8, adapt=false)
+    nthr = length(partition)
+    n = length(x0)
+    t′ = t0
+    t = fill(t′, size(θ0)...)
+    t_old = copy(t)
+    if G === nothing
+        G = [i => rowvals(F.Γ)[nzrange(F.Γ, i)] for i in eachindex(θ0)]
+    end
+    inner = [all(j -> partition(j)[1] == partition(i)[1], js) for  (i,js) in G]
+ #   @show inner
+    G1 = [i => rowvals(F.Γ)[nzrange(F.Γ, i)] for i in eachindex(θ0)]
+
+    @assert all(a.second ⊇ b.second for (a,b) in zip(G, G1))
+
```

Comment:
Important assumption: The independence structure of the bounds `F.Γ` should factorise into partitions. Add a check?